Logistical sorter - single item & muffling upgrade

Zajozor opened the pull request:
Fixes https://github.com/mekanism/Mekanism/issues/4817

I added a button for the logistical sorter, that allows you to toggle between sending single items and sending whole stacks(the way it works now).
Other than that, I added some entries to the lang file and updated the tooltips for buttons to say a little more meaningful messages.
The GUI png file was updated, and I removed the xcf file, which I believe was kept in the repository only by mistake.

What I did not test:
- whether the method `setSingleItem` for CC works properly (only wrote the code analogously to other buttons)
- Obscure situations when the LS gets un-chunkloaded, or the target inventories do, etc. but I guess the change should not introduce new bugs.

### Screenshots
Looks like this:
![image](https://user-images.githubusercontent.com/1835434/57009935-10d68100-6bfa-11e9-8188-6b85bfc15104.png)
Credits for the button icon go to @Sandyna :)

Tested with a setup like this:
![image](https://user-images.githubusercontent.com/1835434/57010006-83dff780-6bfa-11e9-8c26-85836054852a.png)
![image](https://user-images.githubusercontent.com/1835434/57010015-90fce680-6bfa-11e9-90dd-91545404d132.png)

Sending items with single item mode on works pretty well.
![lsorter_in_action](https://user-images.githubusercontent.com/1835434/57010325-5dbb5700-6bfc-11e9-9616-46d7163a7141.gif)


The new tooltips:
![image](https://user-images.githubusercontent.com/1835434/57010129-60697c80-6bfb-11e9-86ad-75956028e1be.png)
![image](https://user-images.githubusercontent.com/1835434/57010133-63fd0380-6bfb-11e9-8c16-f4e7add0a80a.png)
![image](https://user-images.githubusercontent.com/1835434/57010139-68292100-6bfb-11e9-8edb-88bf71b9f249.png)

Zajozor commented:
Also wanted to ask:

1) The code in `TileEntityLogisticalSorter` seems a little messy due to duplicate code between methods (`getNetworkedData` `getGenericPacket` `getConfigurationData``readFromNBT` etc..). If you want, I can refactor it to what I believe would look and read a little nicer.

2) I was also thinking about looking into #5211 , #4813 (this one annoyed me personally too :D ) , #4966 , maybe #4756 , or any other problems regarding logistical sorter if you know of any. Do you feel like all of these would be cool to implement? Any order of priority, or shall I choose? Or perhaps other plans/problems with the LogSorter I could look into?

pupnewfster commented:
At least from the pictures it looks good (I will look into the code and testing it a bit later when I have some more time). Technically it is already possible to do this sort of by setting the stack size of each filter to 1, however I do agree that this is a nice QoL thing to have and makes it even more powerful and easier to use.

In terms of your questions, it is up to you if you make the code more readable, at some point we do plan on rewriting the network/sync code more in general so that it isn't all done through one master packet which should make it easier to not accidentally put things in the wrong order causing the data to be corrupted. Though having the code be easier to read and manage is never a bad thing.

In terms of any priority of what order you want to make PRs that is up to you. It would be nice to have #5435 fixed if you are able to reproduce it easily (I tried some and was unable to), but likely that just has some issue to do with NBT/it grouping them into a single request packet.

Also in terms of if you do look at #4756 I would recommend using JEI's GhostSlot stuff instead of the IRecipeTransferHandler (I think those two are different systems). Basically would do it the same way Ender IO has it, where you can drag items from JEI into the GUI for the filter. Though honestly that may be better to hold off on until after the internals of the filter system get cleaned up as there is a lot of duplicate code between Logistical Sorter and the Digital Miner, and the various types of filters/subfilters they have.

pupnewfster commented on src/main/java/mekanism/client/gui/GuiLogisticalSorter.java:
```diff
@@ -511,11 +522,15 @@ protected void drawGuiContainerForegroundLayer(int mouseX, int mouseY) {
 
         if (xAxis >= 12 && xAxis <= 26 && yAxis >= 110 && yAxis <= 124) {
-            drawHoveringText(LangUtils.localize("gui.autoEject"), xAxis, yAxis);
+            drawHoveringText(LangUtils.localize("gui.logisticalSorter.autoEject.tooltip"), xAxis, yAxis);
```
I think the tooltips (this one and the other two added) should wrap the text rather than just straight up be drawn. This is because some of the descriptions are somewhat long so if you do not have Minecraft in full screen it can run off the edge slightly. (Should be manageable automatically similar to how our item tooltips have their description.)

Zajozor replied:
Good point, will look into it 👍

thiakil replied:
Please also ensure any new language keys are prefixed with `mekanism.`

thiakil commented:
The logistical sorter seems to already have a size based mode, but it's just on the ItemStack Filter 🤔

Zajozor commented:
The tooltips are wrapped, and I prefixed all logistical sorter entries in the lang files as requested.

Good that you mention that this could be achieved using `ItemStack` filters, I actually broke that functionality by calling `getItemAmountFromInventory` instead of `getStackFromInventory` (which is overriden there). I will go through all the filter options thoroughly again, to make sure I did not forget stuff like this.

I feel like the purpose of this button is two things:
- You want to always take one item, while you may have many filters that you do not want to configure manually.
- You do not want to configure any filters -> want to take out any item, but send it one by one (usually with round robin).

Is that enough to have that functionality added, or you rather do not feel like its worth it?

In the next days -> will fix my changes breaking setting item count with item filters, will try to replicate #5435 , and #4813 seems like something I can take on as well.

thiakil commented:
> Is that enough to have that functionality added, or you rather do not feel like its worth it?

Yeah it's fine, if this option is enabled it should override the individual filter settings I think. I'd recommend adding `(overrides filters)` to the tooltip

Zajozor commented:
Sorry for the long time, here are the requested changes (#5435 was fixed in the meantime).

- fixed the code, to work as requested = single item mode overrides ItemStack sizeMode
- did some minor code cleanups
- gui information
When single item mode is turned off, looks like normal:
![image](https://user-images.githubusercontent.com/1835434/57391686-4dd4e180-71bf-11e9-911c-e841a0ec4416.png)

When you turn on single item mode and also have sizeMode turned on:
![image](https://user-images.githubusercontent.com/1835434/57391705-59c0a380-71bf-11e9-8855-0b81698be7c3.png)
With a tooltip (only when singleItem is turned on):
![image](https://user-images.githubusercontent.com/1835434/57391719-5e855780-71bf-11e9-96bf-c4f576cb0280.png)

Also, the singleItem button tooltip tells you it overrides sizeMode:
![image](https://user-images.githubusercontent.com/1835434/57391828-91c7e680-71bf-11e9-8d95-e09d7c33e002.png)


- rebased the branch above your 1.12

Will look into muffling upgrade support as next feature.

Zajozor commented:
Also fixes #4813 

Changes to add muffling upgrade to logistical sorter:
- Add Upgrades tab to the GUI and implement `IUpgradeTile`
- Turn logistical sorter into `TileEntityEffectsBlock`, so that it plays the sound through `SoundHandler` (which respects muffling)
- Add a constructor to `TileEntityEffectsBlock` that allows you to set the `rapidChangeThreshold` (otherwise, the logistical sorter would keep playing the clicking sound for 10 ticks like other machines, but here it results in "more" ticks than we want)
- Add guiID 59, so that when you press the back button from the UpgradesManagement GUI you can get back

Images:
![image](https://user-images.githubusercontent.com/1835434/57405381-92bc4080-71de-11e9-84e1-c2c8290fc568.png)
![image](https://user-images.githubusercontent.com/1835434/57405386-994ab800-71de-11e9-97f3-c6eded8b854c.png)

thiakil commented:
For future, please try to keep separate features in separate PRs (where possible)

thiakil commented on src/main/java/mekanism/client/gui/GuiLogisticalSorter.java:
```diff
@@ -377,13 +382,16 @@ protected void drawGuiContainerForegroundLayer(int mouseX, int mouseY) {
 
-        fontRenderer.drawString("RR:", 12, 74, 0x00CD00);
+        fontRenderer.drawString(LangUtils.localize("mekanism.gui.logisticalSorter.singleItem") + ":", 12, 48, 0x00CD00);
+        fontRenderer.drawString(LangUtils.localize("gui." + (tileEntity.singleItem ? "on" : "off")), 27, 60, 0x00CD00);
```
Would you kindly change these on/off localisations to use `mekanism.common.util.LangUtils#transOnOff`?

Will save some trouble down the track when I fix the lang prefix issues

thiakil commented on src/main/java/mekanism/client/gui/filter/GuiTItemStackFilter.java:
```diff
@@ -128,7 +128,11 @@ protected void drawGuiContainerForegroundLayer(int mouseX, int mouseY) {
         fontRenderer.drawString(LangUtils.localize("gui.itemFilter.max") + ":", 128, 32, 0x404040);
-        fontRenderer.drawString(LangUtils.localize("gui." + (filter.sizeMode ? "on" : "off")), 141, 46, 0x404040);
+        String sizeModeString = LangUtils.localize("gui." + (filter.sizeMode ? "on" : "off"));
+        if (tileEntity.singleItem && filter.sizeMode)
```
Please always use braces.

If you're using IntelliJ IDEA, this can be imported (save it to the project, no overwrite your defaults): https://github.com/mekanism/Mekanism/blob/1.12/docs/GoogleStyle4Indents.xml

Zajozor replied:
I have it imported, yet it showed me nothing wrong with missing braces, could you maybe point me in some direction where i can look to fix this?
![image](https://user-images.githubusercontent.com/1835434/57552008-ccba4d80-736a-11e9-8672-99796e092c91.png)

pupnewfster replied:
On the left under `Code Style` click on the sub tab for `Java`. It should have things like `Force braces` set to `always` under the `Wrapping and Braces` tab for if statements. My guess is that for some reason the scheme didn't load for the java specific scheme.

Zajozor replied:
strange, I do have it turned on. So probably some cache problem at my side, will try clearing it.
![image](https://user-images.githubusercontent.com/1835434/57552935-b9f54800-736d-11e9-97e9-6dd719b233b6.png)
thanks anyway

pupnewfster replied:
Actually thinking about it, probably has to do with it not automatically running. (As in if you didn't put the braces out of habit anyways it wouldn't add them until you run the formatter `CTRL+ALT+L` for current file. For now if you want you can just add the braces manually though given after we get the open PRs merged I am going to do a quick pass making sure things are easier to read across the board since we are changing some of the formatting styling in #5463 .

thiakil commented on src/main/java/mekanism/client/gui/filter/GuiTItemStackFilter.java:
```diff
@@ -151,7 +155,10 @@ protected void drawGuiContainerForegroundLayer(int mouseX, int mouseY) {
         if (xAxis >= 128 && xAxis <= 139 && yAxis >= 44 && yAxis <= 55) {
-            drawHoveringText(LangUtils.localize("gui.sizeMode"), xAxis, yAxis);
+            String sizeModeTooltip = LangUtils.localize("gui.sizeMode");
+            if (tileEntity.singleItem && filter.sizeMode)
```
Braces

thiakil commented on src/main/java/mekanism/common/content/transporter/TItemStackFilter.java:
```diff
@@ -20,18 +20,10 @@
-              : itemType.getItem() == itemStack.getItem());
+        return super.canFilter(itemStack, strict) &&
+                !(strict && sizeMode && (max == 0 || itemStack.getCount() < min)) &&
+                (itemType.getHasSubtypes() ?
```
This subtype check chain (from `itemType.getHasSubtypes() `) should be replaced with `net.minecraftforge.items.ItemHandlerHelper#canItemStacksStackRelaxed`

thiakil commented on src/main/java/mekanism/common/content/transporter/TransporterFilter.java:
```diff
@@ -60,12 +60,18 @@ public static TransporterFilter readFromPacket(ByteBuf dataStream) {
+        return getItemAmountFromInventory(searcher, 64);
+    }
+
+    public InvStack getItemAmountFromInventory(StackSearcher searcher, int amount) {
```
this should have the same method name since one calls the other

thiakil commented on src/main/java/mekanism/common/tile/TileEntityLogisticalSorter.java:
```diff
@@ -95,22 +100,25 @@ public void onUpdate() {
                 for (TransporterFilter filter : filters) {
                     for (StackSearcher search = new StackSearcher(back, facing.getOpposite()); search.getSlotCount() >= 0; ) {
-                        InvStack invStack = filter.getStackFromInventory(search);
+                        InvStack invStack = singleItem ? filter.getItemAmountFromInventory(search, 1)
```
I think this would be clearer to put the ternary as the number param `singleItem ? 1 : 64`

Zajozor replied:
well, this was used because there were two methods - `getStackFromInventory` and getItemAmountFromInventory, only one was overriden in `TItemStackFilter`, so I wanted to call it only in one of the cases.

Will rewrite this to require only one `getStackFromInventory` method and make it cleaner.

thiakil commented:
Functionality checks out

pupnewfster commented on src/main/java/mekanism/common/tile/TileEntityLogisticalSorter.java:
```diff
@@ -46,32 +46,37 @@
     public int rrIndex = 0;
     public int delayTicks;
-    public boolean isActive;
     public boolean clientActive;
```
I haven't tested this stuff yet, but my suspicion is that some stuff between clientActive/isActive is still duplicated now/not needed due to the changing superclass.

Zajozor replied:
Probably right.

Not sure how to test this, as I understand only the surface of this networking stuff, but I removed the `isActive` and `clientActive` logic from `TileEntityLogisticalSorter` and all seems good.

pupnewfster replied:
Added back one small part you removed that made the sorter update so the light on the side would turn on. Given it turns out the `TileEntityEffectsBlock` doesn't call it quite often enough in specific cases at least with the sorter which should have it happen more often.

Zajozor commented:
So I did incorporate all the requested changes:
- Using `Langutils#transOnOff`
- Code style changes (braces, simpler method structure in `getStackFromInventory`
- I removed the `isActive` and `clientActive` logic from `TileEntityLogisticalSorter` which seemed duplicate to me as well, but I am not totally certain about this.

Please let me know if there is anything else I should change/improve.
